docs/ABI.rst

```diff
@@ -1160,11 +1160,26 @@ Property behaviors are implemented using private protocol conformances.
   requirement ::= type assoc-type-name 'Rt' GENERIC-PARAM-INDEX     // same-type requirement on associated type
   requirement ::= type assoc-type-list 'RT' GENERIC-PARAM-INDEX     // same-type requirement on associated type at depth
   requirement ::= type substitution 'RS'                            // same-type requirement with substitution
+  requirement ::= type 'Rl' GENERIC-PARAM-INDEX LAYOUT-CONSTRAINT   // layout requirement
+  requirement ::= type assoc-type-name 'Rm' GENERIC-PARAM-INDEX LAYOUT-CONSTRAINT    // layout requirement on associated type
+  requirement ::= type assoc-type-list 'RM' GENERIC-PARAM-INDEX LAYOUT-CONSTRAINT    // layout requirement on associated type at depth
+  requirement ::= type substitution 'RM' LAYOUT-CONSTRAINT                           // layout requirement with substitution
 
   GENERIC-PARAM-INDEX ::= 'z'                // depth = 0,   idx = 0
   GENERIC-PARAM-INDEX ::= INDEX              // depth = 0,   idx = N+1
   GENERIC-PARAM-INDEX ::= 'd' INDEX INDEX    // depth = M+1, idx = N
 
+  LAYOUT-CONSTRAINT ::= 'N'  // NativeRefCountedObject 
+  LAYOUT-CONSTRAINT ::= 'R'  // RefCountedObject 
+  LAYOUT-CONSTRAINT ::= 'T'  // Trivial 
+  LAYOUT-CONSTRAINT ::= 'E' SIZE-AND-ALIGNMENT  // Trivial of exact size 
+  LAYOUT-CONSTRAINT ::= 'M' SIZE-AND-ALIGNMENT  // Trivial of size at most N bits 
+  LAYOUT-CONSTRAINT ::= 'U'  // Unknown layout
+
+  SIZE-AND-ALIGNMENT ::= INDEX // Size only
+  SIZE-AND-ALIGNMENT ::= INDEX INDEX // Size followed by alignment
```

Comment:
This does not work, because it's not clear where the 'RM' operator ends. The second INDEX could be a second index or a following identifier (which starts with the length).

I suggest using different LAYOUT-CONSTRAINT letters, e.g. 'E' + 'e' and 'M' + 'm' to distinguish between a single and double INDEX.